### PR TITLE
Standardize str argument verification in datasets

### PR DESCRIPTION
**Pull request by pmeier**
First steps towards #1132. I left some `TODO`s and `FIXME` for review / discussion.

### CODE REVIEW

File: torchvision/datasets/utils.py [hunk 1]
```diff
@@ -249,3 +250,16 @@ def download_and_extract_archive(url, download_root, extract_root=None, filename
...
     print("Extracting {} to {}".format(archive, extract_root))
     extract_archive(archive, extract_root, remove_finished)
+
+
+def verify_str_arg(value, valid_values, arg):
+    if not isinstance(value, torch._six.string_classes):
+        msg = "Expected type str for argument {}, ".format(arg)
+        msg += "but got type {}.".format(type(value))
+        raise ValueError(msg)
+    # TODO: Should we call .lower() on value and valid_values?
```
Comment (fmassa) [inline torchvision/datasets/utils.py]: I think it's better to do exact string comparison, so not calling `.lower()` seems more robust to me.

If users want to have less strict checkings, they can do it on the dataset

File: torchvision/datasets/voc.py [hunk 2]
```diff
@@ -83,7 +83,10 @@ def __init__(self,
         self.url = DATASET_YEAR_DICT[year]['url']
         self.filename = DATASET_YEAR_DICT[year]['filename']
         self.md5 = DATASET_YEAR_DICT[year]['md5']
-        self.image_set = image_set
+        # FIXME: Is it common to not have the *.txt of the image_set in the root folder?
+        #  Previous code checked for this, but only reported a wrong value for image_set
```
Comment (fmassa) [inline torchvision/datasets/voc.py]: Not common, and not even supported in the current implementation, so we are good here.

File: torchvision/datasets/svhn.py [hunk 3]
```diff
@@ -43,12 +43,7 @@ def __init__(self, root, split='train', transform=None, target_transform=None,
                  download=False):
         super(SVHN, self).__init__(root, transform=transform,
                                    target_transform=target_transform)
-        self.split = split  # training set or test set or extra set
-
-        if self.split not in self.split_list:
-            raise ValueError('Wrong split entered! Please use split="train" '
-                             'or split="extra" or split="test"')
-
+        self.split = verify_str_arg(split, ("train", "extra", "test"), "split")
```
Comment (fmassa) [inline torchvision/datasets/svhn.py]: I think being explicit is good. But I wonder if it would make sense to just use `self.split_list`, as it was before?

File: torchvision/datasets/stl10.py [hunk 4]
```diff
@@ -167,4 +163,6 @@ def __load_folds(self, folds):
                     list_idx = np.fromstring(str_idx, dtype=np.uint8, sep=' ')
                     self.data, self.labels = self.data[list_idx, :, :, :], self.labels[list_idx]
             else:
+                # FIXME: docstring allows None for folds (it is even the default value)
```
Comment (fmassa) [inline torchvision/datasets/stl10.py]: This would raise an error before, right? Maybe the docstring should be updated?

File: torchvision/datasets/sbd.py [hunk 5]
```diff
@@ -65,11 +65,11 @@ def __init__(self,
 
         super(SBDataset, self).__init__(root, transforms)
 
-        if mode not in ("segmentation", "boundaries"):
-            raise ValueError("Argument mode should be 'segmentation' or 'boundaries'")
-
-        self.image_set = image_set
-        self.mode = mode
+        # FIXME: Is it common to not have the *.txt of the image_set in the root folder?
```
Comment (fmassa) [inline torchvision/datasets/sbd.py]: don't think this is common (and was not even supported before), this looks good to me.

File: torchvision/datasets/lsun.py [hunk 6]
```diff
@@ -75,27 +77,28 @@ def __init__(self, root, classes='train', transform=None, target_transform=None)
...
-                c_short = c.split('_')
-                c_short = c_short.pop(len(c_short) - 1)
-                if c_short not in dset_opts:
-                    raise (ValueError('Unknown postfix: ' + c_short + '.'
-                                      'Options are: ' + str(dset_opts)))
-        else:
-            raise (ValueError('Unknown option for classes'))
-        self.classes = classes
+        except ValueError:
+            # TODO: Should this check for Iterable instead of list?
```
Comment (fmassa) [inline torchvision/datasets/lsun.py]: Probably, but this deserves to be fixed in a separate PR

File: torchvision/datasets/lsun.py [hunk 7]
```diff
@@ -75,27 +77,28 @@ def __init__(self, root, classes='train', transform=None, target_transform=None)
...
+            if isinstance(classes, list):
+                for c in classes:
+                    # TODO: This assumes each item is a str (or subclass). Should this
```
Comment (fmassa) [inline torchvision/datasets/lsun.py]: sure, why not? Maybe in a follow-up PR?

File: torchvision/datasets/lsun.py [hunk 8]
```diff
@@ -75,27 +77,28 @@ def __init__(self, root, classes='train', transform=None, target_transform=None)
...
+                    #   also be checked?
+                    c_short = c.split('_')
+                    category, dset_opt = '_'.join(c_short[:-1]), c_short[-1]
+                    # TODO: "LSUN class" and "postfix" are not arguments.
+                    #  verify_str_arg needs to be adapted to reflect this
```
Comment (fmassa) [inline torchvision/datasets/lsun.py]: I'm not sure I follow, can you explain a bit more?

File: torchvision/datasets/cityscapes.py [hunk 9]
```diff
@@ -109,22 +109,19 @@ def __init__(self, root, split='train', mode='fine', target_type='instance',
...
-
-        if mode == 'fine' and split not in ['train', 'test', 'val']:
-            raise ValueError('Invalid split for mode "fine"! Please use split="train", split="test"'
-                             ' or split="val"')
-        elif mode == 'coarse' and split not in ['train', 'train_extra', 'val']:
-            raise ValueError('Invalid split for mode "coarse"! Please use split="train", split="train_extra"'
-                             ' or split="val"')
+        verify_str_arg(mode, ("fine", "coarse"), "mode")
+        if mode == "fine":
+            # TODO: add condition "invalid argument of mode == fine ..."
```
Comment (fmassa) [inline torchvision/datasets/cityscapes.py]: maybe make `msg` an optional argument in `verify_str_arg`, which defaults to the current message if not passed?

File: torchvision/datasets/utils.py (hunk 1 above)
Comment (pmeier) [reply torchvision/datasets/utils.py]: Should I keep it in `CelebA` for BC?

https://github.com/pytorch/vision/blob/d9830d8688bc2f2fc8070a4688d590c7c0686b27/torchvision/datasets/celeba.py#L69-L75

File: torchvision/datasets/lsun.py (hunk 8 above)
Comment (pmeier) [reply torchvision/datasets/lsun.py]: The error would read: "Unknown value {} for argument LSUN class.". I think this might be confusing, since LSUN class is not an argument. 

I would solve this with your proposal from below: overriding the default message if custom is given.

Comment (fmassa) [reply torchvision/datasets/lsun.py]: sounds good

File: torchvision/datasets/utils.py (hunk 1 above)
Comment (fmassa) [reply torchvision/datasets/utils.py]: Yeah, just add a `split.lower()` in `celeba.py`

**Comment by fmassa** [comment]
Test failure is unrelated and `test_randomresized_params` seems flaky, I opened an issue in https://github.com/pytorch/vision/issues/1170

**Comment by pmeier** [comment]
Should be done now.

> Also, having tests for more datasets would make reviewing dataset PRs a lot easier!

Agreed. If I find some time, I try to add some more.
